src/VisualStudio/Core/Test.Next/Services/ServiceHubServicesTests.cs

```diff
@@ -321,6 +327,344 @@ public async Task TestRemoteWorkspaceCircularReferences()
             Assert.NotNull(solution);
         }
 
+        private static ImmutableArray<ImmutableArray<T>> Permute<T>(T[] values)
```

Comment:
the tests are written as a series of generate document *steps*.  in other words, they say "in step 1, we generate these X docs, then in step 2 these Y docs, and so on and so forth".  To really hammer this system, we then produce every permtuation of that list (e.g. "in step *1* we generate these *Y* docs, in step *2* we generate these *X* docs") to help cover a wide gamut of cases.